### DISCUSSION TITLE
Problema con ciclo for kata 4

### DISCUSSION THREAD

**Discussion by RaulRU30** [LaunchX-InnovaccionVirtual/CursoIntroPython#76] [Q&A] [answered]
Que tal, alguien me podria echar la mano? No logro estructurar bien el ciclo for para imprimir solo las cadenas que contengan las palabras de la lista, vengo de programar en C pero no me termino de entender la sintaxis de los ciclos for y el manejo de listas en Python :c 
![image](https://user-images.githubusercontent.com/49573958/153351244-abe9cb94-b492-4f4b-ad20-cf9c03587245.png)

**Comment by Dfredude** [comment, marked as answer]
Hay muchoos detalles en tu codigo, veamos.
• El metodo `split()` no modifica el **_String_** `text` si no que regresa un nuevo **_List_** de **_Strings_** el cual lo tienes que asignar a una variable para poder usarlo. Tal ejemplo seria: 
`nuevo_text = text.split()` o incluso `text = text.split()`
Practicamente esto soluciona un 50% de los errores 😎
• Tip: Usa variables descriptivas, hace que te equivoques menos en tu propio codigo, asi como se vuelve mas entendible para otras personas. Por ejemplo `for i in text:` podria ser `for sentence in text` para saber que la variable es una oración.
• El metodo `find()` no funciona en **_Lists_**, de hecho `text` ni siquiera deberia de estar en la linea **4**, ya estas iterando en el!
Entonces `find()` se usaria correctamente con `i.find(j)` ya que `i` es la oracion y `j` la palabra que quieres buscar dentro de la oracion.
Aparte este metodo regresa el **_indice_** si lo encuentra o -1 si no lo encuentra. Entonces tambien tendrias que actualizarlo (con variables inventadas para ser descriptivo) :
```
if enunciado.find(palabra) != -1:
    print(enunciado)
```


Esos problemas son los mas importantes, espero haberte ayudado!

**Comment by doguedogue** [comment]
1. Ibas bien, solo que el texto que vas a buscar si lo contiene la frase es j (puedes usar find o in)
2. Te aconsejo poner impresiones de consola para ver que está pasando

```
for i in text:
    print("Sentencia: "+i)
    for j in key_words:
        print("Palabra clave: "+j)
        if i.find(j):
            print("Encontrada: " + i)
            break
```

![image](https://user-images.githubusercontent.com/23409026/153363911-ed545098-31b9-46b3-93d3-6ae86d1d56ae.png)

**Comment by JorgeEHE** [comment]
Hola @RaulRU30 ,

trataré de explicarlo a como lo entiendo yo, ya que al igual que tu yo programaba en **C** (muchos años ya de eso, por eso estoy aquí retomando esos viejos vicios) y al estar ahora en **python** también me costó un poco de trabajo al inicio entender como son las asignaciones de datos al usar el ciclo **for** . Espero con esta imagen quede claro, para ello tomé la parte de tu código donde marca el error y espero ejemplificar el porqué:

![image](https://user-images.githubusercontent.com/99061430/153384695-b050b77a-1b26-44d1-9ae1-26d0f4b6438c.png)

**Nota:** No recuerdo si .find() es realmente un operador o un método, @FernandaOchoa me puedes ayudar con eso para no estar escribiendo cosas que no son?  ;D

**Comment by nosoyhector** [comment]
En Python no existe una implementación ```for``` similar a C, el funcionamiento de este ```for``` es recorrer la colección de objetos uno por uno. Funciona similar al ```foreach``` de PHP y el ```each()``` de jQuery.

La estructura simple de un ```for``` en python es:
```
for <var> in <iterable>:
    <statement(s)>
```
donde ```<iterable>``` es tu colección de objetos, y ```<var>``` es la variable a la que se asigna un objeto a la vez de tu colección de objetos en el ciclo.

El ciclo consiste en:
1. obtiene el primer elemento de tu colección ```<iterable>``` y lo asigna a la variable ```<var>```
2. ejecuta tu(s) declaracio(nes) ```<statement(s)>```
3. si ```<iterable>``` contiene más de un elemento continua, sino termina
4. se asigna a la variable ```<var>``` el siguiente elemento de tu coleccion ```<iterable>```
5. ejecuta tu(s) declaracio(nes) ```<statement(s)>```
6. si ```<iterable>``` contiene más elementos aún por recorrer continua al paso 4, sino termina

También puedes terminar un ciclo con la sentencia ```break```
